fastchat/train/llama2_flash_attn_monkey_patch.py

```diff
@@ -0,0 +1,259 @@
+    )
+    # shape: (b, s, num_heads, head_dim)
+
+    kv_seq_len = k.shape[1]
+    past_kv_len = 0
+    if past_key_value is not None:
+        past_kv_len = past_key_value[0].shape[1]
+        kv_seq_len += past_kv_len
+
+    cos_sin = self.rotary_emb(v, seq_len=kv_seq_len)
+    q, k = apply_rotary_pos_emb(q, k, cos_sin, position_ids)
+
+    if past_key_value is not None:
+        # reuse k, v
+        k = torch.cat([past_key_value[0], k], dim=1)
+        v = torch.cat([past_key_value[1], v], dim=1)
+
+    past_key_value = (k, v) if use_cache else None
+
+    key_padding_mask = attention_mask
+    # Ideally we could just do this:
+    #  q, indices, cu_q_lens, max_s = unpad_input(q, key_padding_mask[:, -q_len:])
+    # but this does not work as Flash attention treats the q seq and kv seq as starting at index 0
+    # which then breaks the causality logic. Probably if q_len >> past_kv_len we should
+    # just skip flash attention. Leaving this in for now to demonstrate correctness of
+    # flash attention information even when q needs padding.
+    # TODO(siddartha): delegate back to original implementation on this condition.
+    if past_kv_len > 0:
+        q = torch.cat(
+            (
+                torch.full(
+                    (bsz, past_kv_len, self.num_heads, self.head_dim),
+                    0.0,
+                    dtype=q.dtype,
+                    device=q.device,
+                ),
+                q,
+            ),
+            dim=1,
+        )
+
+    if key_padding_mask is None:
+        output = flash_attn_func(q, k, v, 0.0, softmax_scale=None, causal=True).view(
+            bsz, q_len + past_kv_len, -1
+        )
+    else:
+        q, indices, cu_q_lens, max_s = unpad_input(q, key_padding_mask)
+        # We can skip concat and call unpad twice but seems better to call unpad only once.
+        kv, _, cu_k_lens, max_k = unpad_input(
+            torch.stack((k, v), dim=2), key_padding_mask
+        )
+        output_unpad = flash_attn_varlen_kvpacked_func(
+            q,
+            kv,
+            cu_q_lens,
+            cu_k_lens,
+            max_s,
+            max_k,
+            0.0,
+            softmax_scale=None,
+            causal=True,
+        )
+        output_unpad = output_unpad.reshape(-1, self.num_heads * self.head_dim)
+        output = pad_input(output_unpad, indices, bsz, q_len + past_kv_len)
+
+    # Need to strip off the zero query outputs.
+    if past_kv_len > 0:
+        output = output[:, past_kv_len:, ...]
+
+    return self.o_proj(output), None, past_key_value
+
+
+# Disable the transformation of the attention mask in LlamaModel as flash attention
+# takes a boolean key_padding_mask. Fills in the past kv length for use in forward.
+def _prepare_decoder_attention_mask(
+    self, attention_mask, input_shape, inputs_embeds, past_key_values_length
+):
+    # [bsz, seq_len]
+    if past_key_values_length > 0 and attention_mask is not None:
+        attention_mask = torch.cat(
```

Comment:
Is this concatenation necessary? attention_mask passed in should have shape of (bsz, kv_len), and past_kv_len is already in kv_len